[pjanthony2001] iP

pjanthony2001 opened the pull request:
# HAL9000 - Task Tracking Chatbot

_HAL9000_ is your trusty companion for keeping track of your basic tasks. Whether it's managing your to-do list, marking deadlines, adding events, or reminding you of important tasks, HAL9000 is here to assist you. [Visit HAL9000's GitHub](https://github.com/pjanthony2001/ip) for more information.


## Key Features:
1. **Local Storage**: HAL9000 has persistent storage that keeps track of your data between sessions.
2. **Adding Deadlines**: Set due dates for your tasks, and HAL9000 will ensure you never miss them.
3. **Adding Events**: Plan and organize events seamlessly within HAL9000.
4. **Adding Todos**: Create a list of to-dos and manage them conveniently.
5. **Marking/Unmarking Tasks**: Easily mark tasks as completed or incomplete.
6. **Simple Interface**: HAL9000 offers a user-friendly chat-based interface.
7. **Intuitive Commands**: Interact with HAL9000 using natural language commands.

## What's Underneath
Here is a random block of code from the source code `LocalStorage` class:

```
public ArrayList<String> load() {
        ArrayList<String> list = new ArrayList<>();
        while (scan.hasNextLine()) {
            String encodedTask = scan.nextLine();
            list.add(encodedTask);
        }
        return list;
    }
```
WOW HAL9000 is written so well in JAVA!
- Uses Gradle to make things easier
- Follows coding conventions
- Has wonderful JavaDoc comments
- ~~CLEAN~~ SUPER CLEAN to read

## What are you waiting for?

NEVER USE BAD LISTS AGAIN 😵‍💫 :
- [ ]  Complete the project proposal.
- [ ]  Attend Team Meeting on 2024-02-10 at 09:00 AM.
- [ ]  Buy groceries for the week.
- [ ]  Call Mom to wish her a happy birthday.

> "I'm sorry, Dave, I'm afraid I can't do that." - HAL9000

Feel free to ask HAL9000 for assistance with any of your task-related needs!

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

pjanthony2001 committed:
Change the ASCII message when Duke runs to HAL9000

pjanthony2001 committed:
Created UserInterface class, to handle all the logic of the Level 0 requirements

pjanthony2001 committed:
Changed .gitignore to ignore *.class files

Updated the UserInterface to handle Level-1 requirements

pjanthony2001 committed:
Revert "Changed .gitignore to ignore *.class files"

This reverts commit 3d374dce77bd810299c1e39f84bb8ffbccdfa55d.

pjanthony2001 committed:
Called UserInterface in Duke.java to start the polling

Finalised Level-0

pjanthony2001 committed:
Updated UserInterface to fulfill Level-1 requirements

Updated .gitignore to ignore .class files

pjanthony2001 committed:
Updated UserInterface to fulfill Level-2 requrements

pjanthony2001 committed:
Updated UserInterface to handle Level-3 requirements

Created Task class

pjanthony2001 committed:
Added Deadline, Event, Todo that extends Task

Made Task abstract

Created CommandProccessor to process commands

Updated UserInterface to handle Level-4 requirements

pjanthony2001 committed:
Added InputException, ProcessingException to handle Exceptions

Modified CommandProcessor and UserInterface to fulfil Level-5 requirements

pjanthony2001 committed:
Remedied a redundant "throws ProcessingException"

pjanthony2001 committed:
Changed CommandProcessor to handle delete command

Added ArrayList<> instead of Task[]

Added a new message to pop when list is empty

Fulfilled Level 6 requirements

pjanthony2001 committed:
Edited EXPECTED.txt and input.txt for testing

pjanthony2001 committed:
Added Storage encapsulating TempStorage and LocalStorage to handle storage on the txt file and in the ArrayList<Task>
Added Codec to decode and encode from Task to String and vice versa
Added encode method to Tasks that encode tasks to strings for storage
Added identifier field to tasks for encoding/decoding and a new constructor that takes in a boolean "done" as well
Added StartUpException to handle issues when the program loads up
Updated ProcessingException to handle causes as well

Level 7 requirements fulfilled

pjanthony2001 committed:
Merge branch 'branch-level-7'

* branch-level-7:
  Added Storage encapsulating TempStorage and LocalStorage to handle storage on the txt file and in the ArrayList<Task> Added Codec to decode and encode from Task to String and vice versa Added encode method to Tasks that encode tasks to strings for storage Added identifier field to tasks for encoding/decoding and a new constructor that takes in a boolean "done" as well Added StartUpException to handle issues when the program loads up Updated ProcessingException to handle causes as well

pjanthony2001 committed:
Added TimeProcessor class to handle processing and parsing of time
Updated Deadline and Event with the LocalDateTime fields as required
Updated InputException to be constructed with a throwable cause as well
Fixed catch-all exception handling

pjanthony2001 committed:
Merge branch 'branch-level-8'

* branch-level-8:
  Added TimeProcessor class to handle processing and parsing of time Updated Deadline and Event with the LocalDateTime fields as required Updated InputException to be constructed with a throwable cause as well Fixed catch-all exception handling

pjanthony2001 committed:
Added Enums in the form of Command
Updated UserInterface, CommandProcessor to use Command
Updated Event to fix bug when decoding from .txt file
Updated Codec to remove redundant lines and utilise ProcessingExceptions constructor with case
Added new method exceptionCommandProcessing to make the code neater

Level 8 requirements fulfilled

pjanthony2001 committed:
Merge remote-tracking branch 'origin/add-gradle-support'

pjanthony2001 committed:
Added Packaging into ui, command, tasks, exceptions, codec, storage

pjanthony2001 committed:
Added Gradle functionality
Added JavaDoc to all files
Made changes to adhere to CheckStyle
Made changes to access modifiers

pjanthony2001 committed:
Removed vscode settings folder

pjanthony2001 committed:
Add Level-9 requirements

Added Search class add search functionality

Updated CommandProcessor, Storage, TempStorage to reflect changes

Updated Task and subclasses to have a getter for the taskName

pjanthony2001 committed:
Merge branch 'branch-level-9'

* branch-level-9:
  Add Level-9 requirements

pjanthony2001 committed:
Add Unit Testing

Added Unit Testing with JUnit for the following classes:
* TimeProcessor
* Task
* Deadline
* Event
* Todo

Tested JAR file capability with ShadowJar and it works fine

pjanthony2001 committed:
Rename constants to follow coding standard

Renamed the constants in the following classes to reflect the coding
standard:

* Codec
* TimeProcessor

Updated UserInterface to configure the Scanner object in the try block

pjanthony2001 committed:
Add JavaDoc to remaining classes

There is room to improve with the exception classes, and how it was
implemented and handled in the code. I added a new class HalException
to be the parent class for the rest of the new exceptions that I defined

Added JavaDoc to the following classes:
* HalException

Added a close() method to UserInterface and all encapsulated classes:
* CommandProcessor
* Storage
* LocalStorage

Shifted method processCommand into the enum Commmand class

pjanthony2001 committed:
Merge branch 'branch-A-JavaDoc'

* branch-A-JavaDoc:
  Add JavaDoc to remaining classes

pjanthony2001 committed:
Follow Coding Standard for remaining classes

Used check style to ensure that all the classes were in line with the
coding standard, along with the use of exceptions. In this case,
TempStorage had issues with catch-all exception handling that was
corrected

ProcessingException class has a new method added:
* exceptionCommandExecution

InputExcpetion class has the following method refactored:
* exceptionCommandProcessing is now exceptionCommandParsing

pjanthony2001 committed:
Merge branch 'branch-A-CodingStandard'

* branch-A-CodingStandard:
  Follow Coding Standard for remaining classes

pjanthony2001 committed:
Update functionality of mark and unmark

The mark and unmark commands will now show a message if the user is
trying to mark (or unmark) a task that is already marked (or unmarked)

Updated InputException, Processing Exception to always end with a
newline when messages are printed

Updated .gitignore to ignore .txt files generated by HAL9000

pjanthony2001 committed:
Add GUI (placeholder)

The barebones of the GUI with no logic was added, and placeholder
classes were used to make things run. This commit was more to learn
the functionalities of JavaFX

A new package Graphics was added with the following classes:
* Graphics
* DialogBox
* Launcher
* MainWindow
* MyClass

Resources added:
* dave.png
* hal9000.png
* DialogBox.fxml
* MainWindow.fxml

pjanthony2001 committed:
Update GUI

Edited DialogBox.fxml to improve the CSS and style of the bot

Updated placeholder class to change behavior

Updated DialogBox class by making the flip() method more
streamlined.

All that is left for this level is to tidy up the code, and replace
the placeholder class with the actual Duke UI

pjanthony2001 committed:
Change UserInterface to return strings instead of printing to CLI

The following classes were changed to return String instead of printing
to CLI:
* CommandProcessor
* Storage
* TempStorage
* UserInterface

GUI was updated, with the following classes changed:
* Graphics
* MainWindow

Task::toString was updated so that the whitespace to represent when an
task was unmarked, was a bit larger and more apparent

pjanthony2001 committed:
Integrate GUI and Logic

Created a new class to replace the old UserInterface class. This is to
push functionality to the GraphicalUI class (formerly Graphics) and to
encapsulate all the logic in the Logic class.

There are still minor bug fixes to be made with the responses from the
chatbot, but the main functionality is there.

Tests were updated to reflect the change in the Task::toString method
as seen in a previous commit.

Duke class is now the main entry point of the application, and replaces
the Launcher class.

pjanthony2001 committed:
Merge branch 'branch-Level-10'

* branch-Level-10:
  Integrate GUI and Logic
  Change UserInterface to return strings instead of printing to CLI
  Update GUI
  Add GUI (placeholder)

pjanthony2001 committed:
Update save file and data directory

Updated the storage of the save data, to have its own directory and the
following classes were modified for the above change:
* Storage
* LocalStorage
* Logic

Exceptions were fixed so the error messages were nicely formatted:
* InputException
* ProcessingException

Ella-e commented on src/main/java/duke/logic/Logic.java:
```diff
@@ -0,0 +1,77 @@
+
+    private Scanner scan;
+    private CommandProcessor cmd;
+    private boolean startUpSuccess = false;
```
Would it be better to use a more boolean variable name like 'isStartUpSuccess'?

pjanthony2001 replied:
Yep! Thanks for catching it

Ella-e commented on src/main/java/duke/command/CommandProcessor.java:
```diff
@@ -0,0 +1,188 @@
+     */
+    public String processData(Command command, String input) throws InputException, ProcessingException {
+        String result = "";
+        switch (command) {
+
+        case DELETE:
+            result = storage.delete(processDelete(input));
+            break;
```
Good to see using the right indentation for the switch/case block.

pjanthony2001 replied:
Thanks!

Ella-e commented on src/main/java/duke/storage/LocalStorage.java:
```diff
@@ -0,0 +1,92 @@
+package duke.storage;
+
+
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Scanner;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
```
It is good to see the importing of packages are separated instead of using a single '*'.

pjanthony2001 replied:
Thanks!

Ella-e commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,97 @@
+                ArrayList<String> loadList = local.load();
+                temp.load(loadList); //Load from local into temp
+            }
+
```
Maybe there is an extra empty line to delete? I noticed the same issue in several other places too.

pjanthony2001 replied:
Yep, thanks for catching that!

Ella-e reviewed (commented):
Good to see all codes are following the coding standard.

pjanthony2001 committed:
Update according to Peer Reviews

Fixed trailing newlines in Command class
Fixed boolean variable naming in Logic class

Fixed small formatting errors in JavaDocs for the following classes:
* HalException
* ProcessingException
* StartUpException

pjanthony2001 committed:
Add assertions

Changed JavaFX versions to be compatible, .fxml files affected:
* MainWindow.fxml
* DialogBox.fxml

Added logic to Command class to handle not commands that were not
lower case

Added assertions to CommandProcessor::processCommand to ensure that the
Command does not fall through the switch case

Added assertions to TempStorage::load to ensure that the list is empty
when the load() function is called

Amended Storage and LocalStorage to allow for the save functionality to
work without the redundant function LocalStorage::clearall

pjanthony2001 committed:
Update code quality

Added guard clauses instead of if-else statements because it improved
readability. Affected the following classes:
* TempStorage

pjanthony2001 committed:
Merge pull request #1 from pjanthony2001/branch-A-Assertions

Merge branch 'branch-A-Assertions'

Changed JavaFX versions to be compatible, .fxml files affected:

MainWindow.fxml
DialogBox.fxml
Added logic to Command class to handle not commands that were not lower case

Added assertions to CommandProcessor::processCommand to ensure that the Command does not fall through the switch case

Added assertions to TempStorage::load to ensure that the list is empty when the load() function is called

Amended Storage and LocalStorage to allow for the save functionality to work without the redundant function LocalStorage::clearall

pjanthony2001 committed:
Merge pull request #2 from pjanthony2001/branch-A-CodeQuality

Update CodeQuality

Added guard clauses instead of if-else statements because it improved readability. Affected the following classes:

* TempStorage

pjanthony2001 committed:
Add CI

Added CI with .yml file given by the website. Removed textuitesting as
it was not applicable.

pjanthony2001 committed:
Add undo extension

The following commands were added:
* REDO
* UNDO

which led to the update of the following classes to handle the new
commands:
*Command
*CommandProcessor

The package history was created with the following classes:
* History
* HistoryManager
* State

All Tasks were made to implement Cloneable and Serializable Interfaces
but the work has not been implemented fully. Deadline, Event still need
to be done properly.

Storage and Temp storage were updated to facilitate the undo and redo
commands but the actual functionality has not been done fully.

pjanthony2001 committed:
Complete undo extension

Completed undo extension by integrating a start state that is set at
start up. The following classes were modified:
* History
* HistoryManager
* State
There was a bug where objects in the state were still pointing to the
same values, so cloning objects was not a deepcopy. This was fixed by
using the clone method in TempStorage::restore as well

Removed Cloneable interface and methods in Task and Todo.

pjanthony2001 committed:
Merge branch 'branch-BCD-Extension'

* branch-BCD-Extension:
  Complete undo extension
  Add undo extension

pjanthony2001 committed:
Add Icon and Title

Added Icon and Title to GUI

Removed System.out::println used for debugging from the following
classes:
* HistoryManager
* TempStorage
* Command Processor

pjanthony2001 committed:
Fix bug in undoing unmark command

Fixed a bug in undoing unmark command where undo failed to ignore unmark
commands that were unmarking a task that was already unmarked. The
following classes were changed:
* TempStorage

Removed redundant methods and made some fields final in these classes:
* Task
* History

pjanthony2001 committed:
Update GUI

Updated GUI to have background. Classes changed:
* MainWindow
* MainWindow.fxml
* background.png
* application.css

Updated HistoryManager to remove the history file entirely as it was
buggy. Files changed:
* Logic
* CommandProcessor
* HistoryManager

Renamed boolean field for Task

pjanthony2001 committed:
Setup Website

Setup the website and include the screenshot

pjanthony2001 committed:
Update UG

Updated the user guide in the README.md

JavaDocs were updated for the remaining files.

pjanthony2001 committed:
Update UG

pjanthony2001 committed:
Refactor code

Refactored main package to be named hal instead of duke

Updated UG

pjanthony2001 committed:
Add more JUnit testing

Added more JUnit testing in the form of the CodecTest class

pjanthony2001 committed:
Update UG

Added parts to the UG that made command descriptions and use cases
much more specific

pjanthony2001 committed:
Add more testing

Polished up and finished JUnit testing for the following classes:
* Codec
* Deadline
* Event
* Task
* Todo

pjanthony2001 committed:
Update UG

Updated UG to fix some formatting errors

pjanthony2001 committed:
Add Final Updates

Bug fix for empty task name

Removed unnecessary whitespace in switch case in CommandProcessor

Instead of printing stack trace, only prints the error message to the
terminal

Updated UG

chin-herng commented on src/main/java/hal/codec/Codec.java:
```diff
@@ -0,0 +1,127 @@
+     * @return A string representation of the encoded task.
+     */
+    public String encode(Task task) {
+        String[] encodedTask = task.encode();
```
Since this is a multi-valued variable, would it be better to use plural form here to emphasise that it is a collection?

chin-herng commented on src/main/java/hal/command/Command.java:
```diff
@@ -0,0 +1,77 @@
+     * @return The corresponding `Command` enum value.
+     * @throws InputException If there is an issue processing the input or if the command is unrecognized.
+     */
+    public static Command processCommand(String input) throws InputException {
```
Perhaps this method can be named `parseCommand` instead? I feel like the word 'process' does not explain the method accurately enough. I noticed the same issue in several other places too, and I think that replacing 'process' with 'parse' might be more sensible.

chin-herng commented on src/main/java/hal/exceptions/InputException.java:
```diff
@@ -0,0 +1,31 @@
+     * @param cause The cause of the exception (a Throwable object).
+     * @return An `InputException` with a message describing the error during command parsing.
+     */
+    public static InputException exceptionCommandParsing(Command command, String input, Throwable cause) {
```
Would it be better to use verbs for the name of this method? I'm thinking something like `createInputException`. Another option might be simply `of`, though that might violate the given coding standard. I noticed the same issue in several other places too.

chin-herng commented on src/main/java/hal/exceptions/StartUpException.java:
```diff
@@ -0,0 +1,10 @@
+/**
+ * The `StartUpException` class represents an exception related to startup issues in the Duke application.
+ */
+public class StartUpException extends HalException {
```
Might be a bit funny but I find the name misleading since I related to the entrepreneurial startup instead of the process of starting up the application. Would it be better to use `launchException` instead (assuming that you're not a rocket science enthusiast)?

chin-herng reviewed (commented):
Overall, I find your code well-written with little coding standard violations, to the extent where I have to resort to catching nits. I am also well-aware that this PR review is 3-week late :(
